modules/packages/Sort.chpl

```diff
@@ -1216,9 +1218,9 @@ module ShellSort {
     // and see Marcin Ciura - Best Increments for the Average Case of Shellsort
     // for the choice of these increments.
     var n = 1 + end - start;
-    var js,hs:int;
+    var js,hs:idxType;
     var v,tmp:Data.eltType;
-    const incs = (701, 301, 132, 57, 23, 10, 4, 1);
+    const incs = (701, 301, 132, 57, 23, 10, 4, 1):(8*idxType);
     for h in incs {
       hs = h + start;
```

Comment:
And actually, even apart from this PR, if `start` is within 701 of `max(hs.type)`, `hs = h + start` will overflow, and the `v = Data[is]` below will access the array at a lower index than `start`, which may be out of bounds.  If it's not out of bounds, I think it may modify the array outside of `[start..end]`.

e.g., this looks legal

```chapel
use Sort.ShellSort;

var A : [0x7fffffffffffff70..0x7ffffffffffffff0 ] int;
shellSort(A, start=0x7fffffffffffff80, end=0x7ffffffffffffff0);
```
```
$CHPL_HOME/modules/packages/Sort.chpl:1225: error: halt reached - array index out of bounds
note: index was -9223372036854775235 but array bounds are 9223372036854775664..9223372036854775792
```

I don't know if radix sort's use of `shellSort()` can end up doing that, or only direct calls.

(Not related to this PR, though making `hs:idxType` means this can happen at less exotic start values, and thus with arrays whose indices do include the result of the overflow.)

Reply:
It seems like shell sort could be adapted to skip the increments that overflow.

Reply:
Though Paul's observation here is independent of my PR, I was hoping to fix it while here.  Making the following change:

```chapel
    for h in incs {
      // guard against cases in which either the 'incs' value or its use        
      // would cause wraparound                                                 
      if (h > 0 && end-start+1 > h) {
```

avoids the OOB accesses, but doesn't result in a sorted array:

```
129 128 127 126 125 124 123 122 121 120 119 118 117 116 115 114 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71 72 73 74 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92 93 94 95 96 97 98 99 100 101 102 103 104 105 106 107 108 109 110 111 112 113
```

Shellsort has always been mysterious enough to me (*) that I'm not sure I should delay this PR further by trying to figure it out, but if the cause is obvious to either of you, let me know.

[* = when I was in high school, my dad gave me the code for it and challenged me to figure out why it worked; I wasn't successful and haven't looked at it again since then]

Reply:
For now, I've added a future to capture Paul's example to this PR instead of trying to fix it.

Reply:
And opened this issue:  https://github.com/chapel-lang/chapel/issues/16925

Reply:
> avoids the OOB accesses, but doesn't result in a sorted array:
> 
> ```
> 129 128 127 126 125 124 123 122 121 120 119 118 117 116 115 114 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71 72 73 74 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92 93 94 95 96 97 98 99 100 101 102 103 104 105 106 107 108 109 110 111 112 113
> ```

Now that it's merged, I did some debugging and found that that is the correct output.  The call to `shellSort()` passes `start=` 16 entries into `A`, so the first 16 entries are left in the initial reverse order.

(I think that was from an earlier test where I was trying to line up the overflow so that the call to sort directly modified elements of the array outside of `start..end`.  So the sort needed to have `start..end` that was a subset of the array, probably something like indices `0x80..0xf0` out of the array over `0x70..0xf0` for a `uint(8)`-indexed array.  Though I can't replicate the calculations to make that be interesting.)

Reply:
@cassella - we spun off that shell sort issue into issue #16925 so comments there would be a better place if you look into it more.

Reply:
@mppf I saw that issue, but the first-16-elements problem wasn't mentioned there.